File: DisasterServer/Maps/Act9.cs
using DisasterServer.Data;
using DisasterServer.Entities;
using DisasterServer.Session;

namespace DisasterServer.Maps;

public class Act9 : Map
{
	public override void Init(Server server)
	{
		Random random = new Random();
		int wallRandom = random.Next(1, 8);
		int addTimeRandom = random.Next(1,128);
		SetTime(server, 130+addTimeRandom);
		switch (wallRandom)
		{
			case 1:
				Spawn(server, new Act9Wall(0, 0, 1025)); // Поталог
				Spawn(server, new Act9Wall(1, 1663, 0)); // Лево
				Spawn(server, new Act9Wall(2, 1663, 0)); // Право
				break;
			case 2:
				Spawn(server, new Act9Wall(0, 0, 1025)); // Поталог
				Spawn(server, new Act9Wall(1, 1663, 0)); // Лево
				break;
			case 3:
				Spawn(server, new Act9Wall(0, 0, 1025)); // Поталог
				Spawn(server, new Act9Wall(2, 1663, 0)); // Право
				break;
			case 4:
				Spawn(server, new Act9Wall(1, 1663, 0)); // Лево
				Spawn(server, new Act9Wall(2, 1663, 0)); // Право
				break;
			case 5:
				Spawn(server, new Act9Wall(0, 0, 1025)); // Поталог
				break;
			case 6:
				Spawn(server, new Act9Wall(1, 1663, 0)); // Лево
				break;
			case 7:
				Spawn(server, new Act9Wall(2, 1663, 0)); // Право
				break;
			case 8:
				//Ничего
				break;
		}
		Terminal.Log($"WallRandom: {wallRandom.ToString()}"); 
		Terminal.Log($"AddTimeRandom: {addTimeRandom.ToString()}");
		base.Init(server);
	}

	public override void Tick(Server server)
	{
		base.Tick(server);
	}

	protected override int GetPlayerOffset(Server server)
	{
		lock (server.Peers)
		{
			return (server.Peers.Count<KeyValuePair<ushort, Peer>>((KeyValuePair<ushort, Peer> e) => !e.Value.Waiting) - 1) * 10;
		}
	}

	protected override int GetRingSpawnCount()
	{
		return 38;
	}

	protected override float GetRingTime()
	{
		return 3f;
	}

	public override bool CanSpawnRedRings()
	{
		return true;
	}
}

File: disasterserver/maps/Act9.c
#include <maps/Act9.h>
#include <entities/Act9Wall.h>
#include <States.h>
#include <stdlib.h>
#include <time.h>

bool act9_init(Server* server)
{
    if (g_config.random_mode)
    {
        // init the random number generator
        srand((unsigned int)time(NULL));

        int wallRandom = (rand() % 8) + 1; // 1-8
        int addTimeRandom = (rand() % 128) + 1; // 1-128
        RAssert(map_time(server, (2.17 + (addTimeRandom / 60.0)) * TICKSPERSEC, 10));
        
        // Spawn walls according to random selection
        switch (wallRandom)
        {
            case 1: // Ceiling + left + Right
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(0, 0, 1025)), sizeof(Act9Wall), NULL));
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(1, 1663, 0)), sizeof(Act9Wall), NULL));
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(2, 1663, 0)), sizeof(Act9Wall), NULL));
                break;
            case 2: // Ceiling + left
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(0, 0, 1025)), sizeof(Act9Wall), NULL));
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(1, 1663, 0)), sizeof(Act9Wall), NULL));
                break;
            case 3: // Ceiling + Right
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(0, 0, 1025)), sizeof(Act9Wall), NULL));
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(2, 1663, 0)), sizeof(Act9Wall), NULL));
                break;
            case 4: // Left + Right
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(1, 1663, 0)), sizeof(Act9Wall), NULL));
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(2, 1663, 0)), sizeof(Act9Wall), NULL));
                break;
            case 5: // only ceiling
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(0, 0, 1025)), sizeof(Act9Wall), NULL));
                break;
            case 6: // only left
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(1, 1663, 0)), sizeof(Act9Wall), NULL));
                break;
            case 7: // only right
                RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(2, 1663, 0)), sizeof(Act9Wall), NULL));
                break;
            case 8: // nothing (no walls)
                break;
        }
    }
    else
    {
        RAssert(map_time(server, 2.17 * TICKSPERSEC, 10));
        RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(0, 0, 1025)), sizeof(Act9Wall), NULL));
        RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(1, 1663, 0)), sizeof(Act9Wall), NULL));
        RAssert(game_spawn(server, (Entity*)&(MakeAct9Wall(2, 1663, 0)), sizeof(Act9Wall), NULL));
    }
    
    RAssert(map_ring(server, 3));
    
    return true;
}